NE753: Docs work for External DNS Operator (#41811)
openshift/openshift-docs: networking/external_dns_operator/nw-control-dns-records-public-managed-zone-gcp.adoc

$ oc get secret gcp-credentials -n kube-system --template='{{$v := index .data "service_account.json"}}{{$v}}' | base64 -d - > decoded-gcloud.json
----

. Export Google credentials by running the following command:
+
[source,terminal]
----
$ export GOOGLE_CREDENTIALS=decoded-gcloud.json
----

. Activate your account by using the following command:
+
[source,terminal]
----
$ gcloud auth activate-service-account  <client_email as per decoded-gcloud.json> --key-file=decoded-gcloud.json
----

. Set your project by running the following command:
+
[source,terminal]
----
$ gcloud config set project <project_id as per decoded-gcloud.json>
----

. Get the routes to check the domain by running the following command:
+
[source,terminal]
----
$ oc get routes --all-namespaces | grep console
----
+
.Example output
[source,terminal]
----
openshift-console          console             console-openshift-console.apps.test.gcp.example.com                       console             https   reencrypt/Redirect     None
openshift-console          downloads           downloads-openshift-console.apps.test.gcp.example.com                     downloads           http    edge/Redirect          None
----

> . Get your zone that installer created for you. Run the following command:

Author: Addressed

> :_content-type: PROCEDURE

Author: Addressed

[id="nw-control-dns-records-public-hosted-zone-gcp_{context}"]
= Creating DNS records on an public managed zone for GCP by using Red Hat External DNS Operator

> 

Author: Addressed

.Procedure

. Export the `kubeconfig` file:
+
[source,terminal]
----
$ export KUBECONFIG=~/<path to>/kubeconfig
----

> . Check the user. The user must have access to the `kube-system` namespace if you don’t have the credentials as you can fetch the credentials from the `kube-system` namespace to use the cloud provider client:

Reviewer: ```suggestion
. Check the user. The user must have access to the `kube-system` namespace. If you don’t have the credentials, you can fetch the credentials from the `kube-system` namespace to use the cloud provider client:
```

Author: Addressed

+
[source,terminal]
----
$ export GOOGLE_CREDENTIALS=decoded-gcloud.json
----

. Activate your account by using the following command:
+
[source,terminal]
----
$ gcloud auth activate-service-account  <client_email as per decoded-gcloud.json> --key-file=decoded-gcloud.json
----

. Set your project by running the following command:
+
[source,terminal]
----
$ gcloud config set project <project_id as per decoded-gcloud.json>
----

. Get the routes to check the domain by running the following command:
+
[source,terminal]
----
$ oc get routes --all-namespaces | grep console
----
+
.Example output
[source,terminal]
----
openshift-console          console             console-openshift-console.apps.qe.gcp.devcluster.openshift.com                       console             https   reencrypt/Redirect     None
openshift-console          downloads           downloads-openshift-console.apps.qe.gcp.devcluster.openshift.com                     downloads           http    edge/Redirect          None
----

. Get your zone that installer created for you. Run the following command:
+
[source,terminal]
----
$ gcloud dns managed-zones list | grep qe.gcp.devcluster.openshift.com

> qe-cvs4g-private-zone qe.gcp.devcluster.openshift.com

Reviewer 2: I'm not sure if we should add these `gcloud` commands in the document?
And maybe better to replace the real QE domain name with `test.gcp.example.com`?
@alebedev87  WDYT?

Reviewer 3: This is what I tried to discuss in WIP doc (cannot find the comment anymore) but didn't get any reply.
The checks of what was done by ExternalDNS are different for AWS and Azure/GCP. In AWS we just send the user to the web UI of the provider while for Azure and GCP we provider client commands to check in details.

I'm fine with either approach: AWS one is short and easy to document, Azure/GCP is longer but gives more details. I let @xenolinux decide which approach (or mix) is better from the reader perspective.

@lihongan : Azure doc was using a QE domain too. I think it's indeed better to use some made up domain name which doesn't show any internal Red hat kitchen.

Author: Addressed. Replaced real QE domain name with `test.gcp.example.com`

Author: @alebedev87 I'm sorry. I didn't get the idea about the approach. Do we need any additional commands in the AWS document?

----

> . Check if the zone only have DNS records either `NS` or `SOA`:

Reviewer: ```suggestion
. Check if the zone only has DNS records  of either `NS` or `SOA`:
```

Author: Addressed

+
[source,terminal]
----
$ gcloud dns record-sets list --zone=qe-cvs4g-private-zone
----

> . Create a `ExternalDNS` CR for `route`. Run the following command:

Reviewer 3: Same comment about YAML used as command.

Author: Addressed

+
[source,yaml]
----
apiVersion: externaldns.olm.openshift.io/v1alpha1
kind: ExternalDNS
metadata:
  name: sample-gcp <1>
spec:
  domains:
    - filterType: Include <2>
      matchType: Exact <3>
      name: qe.gcp.devcluster.openshift.com <4>
  provider:
    type: GCP <5>
  source:
    openshiftRouteOptions: <6>
      routerName: default <7>
    type: OpenShiftRoute <8>
----

<1> The `name` defines the name of External DNS CR.
<2> By default all hosted zones are selected as potential targets. You can include a hosted zone that you need.
<3> The matching of the target zone's domain has to be exact (as opposed to regular expression match).
<4> Specify the exact domain of the zone you want to update. The hostname of the routes must be subdomains of the specified domain.

> <5> The `type` field Publishes records on public hosted zone GCP.

Reviewer 2: Both public and private hosted zones are supported, so the field `defines Google Cloud DNS provider.`

Author: Addressed

<6> You can define an option for route source under `openshiftRouteOptions` field.

> <7> If the source is `openshift-route` then you can pass the ingress controller name. Based on this name external-dns will select the respective router from the route status and map that routerCanonicalHostname to the route host while creating a CNAME record.

Reviewer: ```suggestion
<7> If the source is `openshift-route`, then you can pass the ingress controller name. Based on this name, external-dns will select the respective router from the route status and map that routerCanonicalHostname to the route host while creating a CNAME record.
```

Author: Addressed

> <8> Defines source route, dns records for routes get created on your cloud DNS.

Reviewer 3: Same suggestions as for AWS.

Author: Addressed

> . Check the records created for OCP routes using the following command:

Reviewer: In the published output, the steps following your annotations are indented and the numbering restarts.

Author: Addressed. Steps are rendering correctly now.

+
[source,terminal]
----

> $ gcloud dns record-sets list --zone=qe-cvs4g-private-zone

Reviewer 3: ```suggestion
$ gcloud dns record-sets list --zone=qe-cvs4g-private-zone | grep console
```
Just as for Azure, let's use the console routes, this will simplify the doc.

Author: Addressed

----

. You can create a route with a sample app:
+
[source,terminal]
----
$ oc new-app --docker-image=openshift/hello-openshift -l app=hello-openshift
$ oc expose service/hello-openshift -l app=hello-openshift
----

. Check the record for the `hello-openshift` route:
+
[source,terminal]
----
$ gcloud dns record-sets list --zone=qe-cvs4g-private-zone | grep hello-openshift
----

. To delete the route and app, run the following command:
+
[source,terminal]
----
$ oc delete all -l app=hello-openshift
----

. Check whether the records are absent:
+
[source,terminal]
----
$ gcloud dns record-sets list --zone=qe-cvs4g-private-zone | grep hello-openshift

> ----

Reviewer 3: No need to do all this as we checked for `console` routes before.

Author: Addressed
